rholang/src/main/scala/coop/rchain/rholang/interpreter/rholangADT.scala

```diff
@@ -61,16 +64,54 @@ case class Par(
         that.receives ++ receives,
         that.evals ++ evals,
         that.news ++ news,
-        that.exprs ++ exprs)
+        that.exprs ++ exprs,
+        Some(GPrivate(java.util.UUID.randomUUID.toString))
+    )
+
+  def procCount: Int =
+    sends.length +
+      receives.length +
+      evals.length +
+      news.length +
+      exprs.length
+
+  def isEmpty: Boolean =
+    sends.isEmpty &&
+      receives.isEmpty &&
+      evals.isEmpty &&
+      news.isEmpty &&
+      exprs.isEmpty &&
+      id.isEmpty
+
+  def isIdentifier: Boolean =
+    sends.isEmpty &&
+      receives.isEmpty &&
+      evals.isEmpty &&
+      news.isEmpty &&
+      exprs.isEmpty &&
+      id.nonEmpty
+
+  override def toString: String =
+    if (isEmpty) "Nil"
+    else if (isIdentifier) id.get.p.substring(0, 8)
+    else {
+      val collect = List(sends, receives, evals, news, if (id.isDefined) List(id) else Nil)
+      (for { xs <- collect; if xs.nonEmpty } yield {
+        xs.mkString(" | ")
+      }).mkString(" | ")
+    }
```

Comment:
Please remove the pretty printing portion. It's good work, it just doesn't belong in this PR